src/clients/UBAClient/UBAClientWithRefresh.ts

```diff
@@ -2,163 +2,967 @@ import assert from "assert";
+    const runningBalanceThresholds = ubaConfigForChain.getBalanceTriggerThreshold(flowChain, tokenSymbol);
+    console.log("- Bundle information for flow", {
+      blockRangesContainingFlow,
+      openingBalanceForChain,
+      ubaConfigBaselineFeeCurve: ubaConfigForChain
+        .getBaselineFee(flow.destinationChainId, flow.originChainId)
+        .toString(),
+      upperRunningBalanceThresholds: {
+        target: runningBalanceThresholds.upperBound?.target?.toString(),
+        threshold: runningBalanceThresholds.upperBound?.threshold?.toString(),
+      },
+      lowerRunningBalanceThresholds: {
+        target: runningBalanceThresholds.lowerBound?.target?.toString(),
+        threshold: runningBalanceThresholds.lowerBound?.threshold?.toString(),
+      },
+    });
+
+    // Use the opening balance to compute expected flow fees:
+    const bundleKey = getBundleKeyForBlockRanges(blockRangesContainingFlow);
+    const precedingValidatedFlows = this.validatedFlowsPerBundle[bundleKey][tokenSymbol][flowChain];
+    const lastFlow = precedingValidatedFlows[precedingValidatedFlows.length - 1];
+    const latestRunningBalance = lastFlow?.runningBalance ?? openingBalanceForChain.runningBalance;
+    const latestIncentiveBalance = lastFlow?.incentiveBalance ?? openingBalanceForChain.incentiveBalance;
+    let potentialBalancingFee: BigNumber;
+    if (isUbaInflow(flow)) {
+      ({ balancingFee: potentialBalancingFee } = getDepositFee(
+        flow.amount,
+        latestRunningBalance,
+        latestIncentiveBalance,
+        flowChain,
+        ubaConfigForChain
+      ));
+    } else {
+      ({ balancingFee: potentialBalancingFee } = getRefundFee(
+        flow.amount,
+        latestRunningBalance,
+        latestIncentiveBalance,
+        flowChain,
+        ubaConfigForChain
+      ));
+    }
+
+    // Figure out the running balance so far for this flow's chain. This is based on all already validated flows
+    // for this chain plus the current flow assuming it is valid.
+    // TODO: Could optimize this more by only passing in the latest flow along with latest balance counts
+    // from last flow in preceding.
+    const { runningBalance, incentiveBalance, netRunningBalanceAdjustment } = analog.calculateHistoricalRunningBalance(
+      precedingValidatedFlows
+        .map(({ flow, balancingFee }) => {
+          return {
+            ...flow,
+            incentiveFeePct: balancingFee,
+          };
+        })
+        .concat({
+          ...flow,
+          incentiveFeePct: potentialBalancingFee,
+        }),
+      openingBalanceForChain.runningBalance,
+      openingBalanceForChain.incentiveBalance,
+      flowChain,
+      tokenSymbol,
+      ubaConfigForChain
     );
+
+    // Figure out the LP fee which is based only on the flow's origin and destination chain.
+    const lpFee = computeLpFeeForRefresh(ubaConfigForChain.getBaselineFee(flow.destinationChainId, flow.originChainId));
```

Comment:
Could `computeLpFeeForRefresh` be relocated into this class? It seems very much targeted for it.